From cdd71b9d5b11d8d568fb9bfc1685c4af549e7ff5 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:18:29 -0400
Subject: [fs] nfsd: Ensure that nfsd_create_setattr commits files to stable storage

Message-id: <1410898849-22126-26-git-send-email-bfields@redhat.com>
Patchwork-id: 95704
O-Subject: [Patch RHEL7 025/165] nfsd: Ensure that nfsd_create_setattr commits files to stable storage
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Trond Myklebust <trond.myklebust@primarydata.com>

Since nfsd_create_setattr strips the mode from the struct iattr, it
is quite possible that it will optimise away the call to nfsd_setattr
altogether.
If this is the case, then we never call commit_metadata() on the
newly created file.

Also ensure that both nfsd_setattr() and nfsd_create_setattr() fail
when the call to commit_metadata fails.

Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: 0f3a24b43bf75adf67df188a85594a8f43b9ee93
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/vfs.c b/fs/nfsd/vfs.c
index edae097..746bf81 100644
--- a/fs/nfsd/vfs.c
+++ b/fs/nfsd/vfs.c
@@ -463,7 +463,7 @@ out_put_write_access:
  if (size_change)
   put_write_access(inode);
  if (!err)
-  commit_metadata(fhp);
+  err = commit_metadata(fhp);
 out:
  return err;
 }
@@ -1272,7 +1272,8 @@ nfsd_create_setattr(struct svc_rqst *rqstp, struct svc_fh *resfhp,
   iap->ia_valid &= ~(ATTR_UID|ATTR_GID);
  if (iap->ia_valid)
   return nfsd_setattr(rqstp, resfhp, iap, 0, (time_t)0);
- return 0;
+ /* Callers expect file metadata to be committed here */
+ return commit_metadata(resfhp);
 }
 
 /* HPUX client sometimes creates a file in mode 000, and sets size to 0.
@@ -1404,9 +1405,10 @@ nfsd_create(struct svc_rqst *rqstp, struct svc_fh *fhp,
  err = nfsd_create_setattr(rqstp, resfhp, iap);
 
  /*
-  * nfsd_setattr already committed the child.  Transactional filesystems
-  * had a chance to commit changes for both parent and child
-  * simultaneously making the following commit_metadata a noop.
+  * nfsd_create_setattr already committed the child.  Transactional
+  * filesystems had a chance to commit changes for both parent and
+  * child * simultaneously making the following commit_metadata a
+  * noop.
   */
  err2 = nfserrno(commit_metadata(fhp));
  if (err2)
@@ -1577,7 +1579,8 @@ do_nfsd_create(struct svc_rqst *rqstp, struct svc_fh *fhp,
  err = nfsd_create_setattr(rqstp, resfhp, iap);
 
  /*
-  * nfsd_setattr already committed the child (and possibly also the parent).
+  * nfsd_create_setattr already committed the child
+  * (and possibly also the parent).
   */
  if (!err)
   err = nfserrno(commit_metadata(fhp));
-- 
1.7.1